django/db/models/fields/json.py

```diff
@@ -0,0 +1,483 @@
+import json
+
+from django import forms
+from django.core import checks, exceptions
+from django.db import connection as builtin_connection
+from django.db.models import Func, Value, lookups
+from django.db.models.lookups import (
+    FieldGetDbPrepValueMixin, Lookup, Transform,
+)
+from django.db.utils import NotSupportedError
+from django.utils.translation import gettext_lazy as _
+
+from . import Field, TextField
+from .mixins import CheckFieldDefaultMixin
+
+
+class JSONField(CheckFieldDefaultMixin, Field):
+    description = _('A JSON object')
+    default_error_messages = {
+        'invalid': _("Value must be valid JSON."),
+    }
+    _default_hint = ('dict', '{}')
+
+    def __init__(self, encoder=None, decoder=None, default=dict, *args, **kwargs):
+        self.encoder, self.decoder = encoder, decoder
+        super().__init__(default=default, *args, **kwargs)
+
+    def check(self, **kwargs):
+        errors = super().check(**kwargs)
+        errors.extend(self._check_json_support())
+        return errors
+
+    @classmethod
+    def _check_json_support(cls):
+        errors = []
+        if not builtin_connection.features.supports_json:
+            errors.append(
+                checks.Error(
+                    _('JSONField is not supported by this database backend.'),
+                    obj=cls
+                )
+            )
+        return errors
+
+    def deconstruct(self):
+        name, path, args, kwargs = super().deconstruct()
+        if self.default is dict:
+            del kwargs['default']
+        if self.encoder is not None:
+            kwargs['encoder'] = self.encoder
+        if self.decoder is not None:
+            kwargs['decoder'] = self.decoder
+        return name, path, args, kwargs
+
+    def from_db_value(self, value, expression, connection):
+        if value is None:
+            return value
+        elif connection.vendor == 'oracle' and value == '':
+            return None
+        else:
+            try:
+                return json.loads(value, cls=self.decoder)
+            except json.decoder.JSONDecodeError:
+                return value
+
+    def get_internal_type(self):
+        return 'JSONField'
+
+    def get_prep_value(self, value):
+        if value is None:
+            return value
+        return json.dumps(value, cls=self.encoder)
+
+    def get_transform(self, name):
+        transform = super().get_transform(name)
+        if transform:
+            return transform
+        return KeyTransformFactory(name)
+
+    def select_format(self, compiler, sql, params):
+        if compiler.connection.vendor == 'postgresql':
```

Comment:
```python
… and self.decoder is not None:
```
If not using a custom decoder, this could be skipped to avoid the cast? Would also need an equivalent branch in `from_db_value ()` to skip doing `json.loads()` if I understand correctly.

Reply:
Yeah, but the psycopg2 docs say that "it is an efficient operation that doesn't involve a copy". Personally, I'd like to keep the difference between database backends minimum. Also, while it is unlikely people would do this, casting the value to `text` wouldn't change the field's behavior if for some reason they did `psycopg2.extras.register_default_json` in their code. This can be good or bad, depending on how you look at it.

Reply:
Hmm. By your quote it just seems that we're trading performance for PostgreSQL when not using a custom decoder. This would mean everyone who was using the existing field will suddenly get a subtly different behaviour.

(Disclaimer: I'm a PostgreSQL user and don't like my backend being crippled to satisfy deficiencies in others...)

Reply:
If you look at psycopg2's source code: https://github.com/psycopg/psycopg2/blob/master/lib/_json.py#L158

The module also uses `json.dumps` and `json.loads` to convert JSON values to Python objects and vice versa. I'm pretty sure there's no database magic involved. It's just that if the module detects that it's a `json`/`jsonb` value by `OID`, it will call the functions from Python's `json` module. The data retrieved from the database would still be a string. If we cast it to `text`, it just bypasses psycopg2's checking and then we call `json.loads` in `from_db_value`. I think that's why the docs say "it is an efficient operation ...".

However, if you still prefer adding
```python
and self.decoder is not None:
```

Let me know and I'll do so :smiley:

Reply:
Will have to have a think. It would be nice to avoid that cast unless absolutely necessary.

Reply:
I decided to avoid the cast :wink: